### DISCUSSION TITLE
VPN tunnel traffic optimization.

### DISCUSSION THREAD

**Discussion by ioagr** [chmorgan/sharppcap#465] [Q&A]
Hello,

I have a question that isn't related to SharpPcap, but I didn't know where else to ask and since I'm using SharpPcap in this project, I thought maybe someone could help me here.
I am trying to create a VPN tunnel between a PC (client) and a private network consisting of client, server and bridge.

![image](https://github.com/dotpcap/sharppcap/assets/112566094/f782898d-b087-4b53-bb9d-5a1426412029)

The first attempt worked, but the latency was around 300ms and the speed was 0.6MB/s. Afterwards I excluded the traffic to and from a list of known DNS and created a hash table on the bridge side, where I save the IPs with which the client tries to communicate and send back to the client only the packets from those IPs . This greatly reduced traffic (latency dropped to 70-100ms and speed increased to ~4MB/s).
Now I wonder if I can reduce the traffic even more. I have no experience in networking and I don't know how communication works at the IP level.
So, my question is if I can delete the IPs added in the hash table after a certain time since the client no longer sent packets to the respective IP so that I no longer send the packets received by the bridge from those IPs to client and keep the size of the hash table to a minimum? Or do you have other suggestions?
Thanks in advance for any help!

EDIT:
This are the filters that I pass to WinDivert:

Client: (ip or ipv6) and !loopback and outbound and ip.SrcAddr=="local PC IP" and (ip.DstAddr!="server IP")
Bridge: (ip or ipv6) and !loopback and inbound and ip.DstAddr=="local PC IP" and (ip.SrcAddr!="server IP")

The rest of the IP filtering (list of DNS and the hash table) is done in the software I wrote because WinDivert freezes if I add too many filters.

**Comment by kayoub5** [comment]
Using TUN/TAP should give you better results than using WinDivert, since OS will automatically optimize what really needs to be forwarded.

**Reply by ioagr** [reply]
OK, so I guess I can only use TUN/TAP on the client side, because TUN/TAP works more like a loopback adapter and doesn't receive packets from the outside. I'm right?

**Reply by kayoub5** [reply]
It can receive packets from the outside (if you inject them), it's a bi-directional communication.
